internal/xfn/function_runner.go

```diff
@@ -0,0 +1,283 @@
+// something other than the package manager will reconcile FunctionRevisions.
+// There's no guarantee it will create a Service, or that the endpoint will
+// remain stable across FunctionRevisions.
+//
+// https://github.com/crossplane/crossplane/blob/226b81f/design/one-pager-package-runtime-config.md
+//
+// With this in mind, we attempt to:
+//
+// * Create a connection the first time someone runs a Function.
+// * Cache it so we don't pay the setup cost every time the Function is called.
+// * Verify that it has the correct target every time the Function is called.
+//
+// In the happy path, where a client already exists, this means we'll pay the
+// cost of listing and iterating over FunctionRevisions from cache. The default
+// RevisionHistoryLimit is 1, so for most Functions we'd expect there to be two
+// revisions in the cache (one active, and one previously active).
+func (r *PackagedFunctionRunner) getClientConn(ctx context.Context, name string) (*grpc.ClientConn, error) {
+	log := r.log.WithValues("function", name)
+
+	l := &pkgv1beta1.FunctionRevisionList{}
+	if err := r.client.List(ctx, l, client.MatchingLabels{pkgv1.LabelParentPackage: name}); err != nil {
+		return nil, errors.Wrapf(err, errListFunctionRevisions)
+	}
+
+	var active *pkgv1beta1.FunctionRevision
+	for i := range l.Items {
+		if l.Items[i].GetDesiredState() == pkgv1.PackageRevisionActive {
+			active = &l.Items[i]
+		}
+	}
+	if active == nil {
+		return nil, errors.New(errNoActiveRevisions)
+	}
+
+	if active.Status.Endpoint == "" {
+		return nil, errors.Errorf(errFmtEmptyEndpoint, active.GetName())
+	}
+
+	r.connsMx.RLock()
+	conn, ok := r.conns[name]
+	r.connsMx.RUnlock()
+
+	if ok {
+		// We have a connection for the up-to-date endpoint. Return it.
+		if conn.Target() == active.Status.Endpoint {
+			return conn, nil
+		}
+
+		// This connection is to an old endpoint. We need to close it and create
+		// a new connection. Close only returns an error is if the connection is
+		// already closed or in the process of closing.
+		log.Debug("Closing gRPC client connection with stale target", "old-target", conn.Target(), "new-target", active.Status.Endpoint)
+		_ = conn.Close()
+	}
+
+	// This context is only used for setting up the connection.
+	ctx, cancel := context.WithTimeout(ctx, dialFunctionTimeout)
+	defer cancel()
+
+	conn, err := grpc.DialContext(ctx, active.Status.Endpoint,
+		grpc.WithTransportCredentials(r.creds),
+		grpc.WithDefaultServiceConfig(lbRoundRobin))
+	if err != nil {
+		return nil, errors.Wrapf(err, errFmtDialFunction, active.Status.Endpoint, active.GetName())
+	}
+
+	r.connsMx.Lock()
+	r.conns[name] = conn
+	r.connsMx.Unlock()
+
+	log.Debug("Created new gRPC client connection", "target", active.Status.Endpoint)
+	return conn, nil
+}
+
+// GarbageCollectConnections runs every interval until the supplied context is
+// cancelled. It garbage collects gRPC client connections to Functions that are
+// no longer installed.
```

Comment:
Were you thinking that we could pass the FunctionRunner into the package reconciler and have it delete Function clients when it noticed a Function was deleted? To be honest I didn't think of that. I think one challenge of doing it that way is that the Package controller seems pretty hard to add specialized code to. It wants to reconcile all types of packages (Provider, Configuration, Function) the same way and AFAICT doesn't have any hooks like the package revision controller does. We could refactor it, but that seems likely to be a lot of work.

I did initially explore passing an informer to this runner so that it could watch Functions and create/delete clients on-demand. I moved away from that approach mostly because:

* I wanted to create a client the first time a Function was called, so we could more easily return any errors setting up the client to the caller. (As opposed to reporting something like "error: no client exists for Function foo" because client-setup is done out-of-band.)
* This felt simpler to implement
* This takes a write lock slightly less often (e.g. if there are multiple Function clients to tear down in a 10 min period)

I could see us potentially splitting the difference. We could create clients on-demand but use an informer to delete them as soon as the Function was uninstalled. I suspect what is here is probably good enough though.

Reply:
> I think one challenge of doing it that way is that the Package controller seems pretty hard to add specialized code to. It wants to reconcile all types of packages (Provider, Configuration, Function) the same way and AFAICT doesn't have any hooks like the package revision controller does.

FWIW, this would indeed be a better fit for the deactivation of a Function Revision so hooks on revisions would be a good place.